src/drivers/hygrometer/sht3x/sht3x.cpp

```diff
@@ -0,0 +1,349 @@
+ *
+ * @author Roman Dvorak <dvorakroman@thunderfly.cz>
+ *
+ */
+
+#include "sht3x.h"
+
+SHT3X::SHT3X(const I2CSPIDriverConfig &config) :
+	I2C(config),
+	ModuleParams(nullptr),
+	I2CSPIDriver(config)
+{
+}
+
+uint8_t SHT3X::calc_crc(uint8_t data[2])
+{
+	uint8_t crc = 0xFF;
+
+	for (int i = 0; i < 2; i++) {
+		crc ^= data[i];
+
+		for (uint8_t bit = 8; bit > 0; --bit) {
+			if (crc & 0x80) {
+				crc = (crc << 1) ^ 0x31u;
+
+			} else {
+				crc = (crc << 1);
+			}
+		}
+	}
+
+	return crc;
+}
+
+
+int SHT3X::set_pointer(uint16_t command)
+{
+	if (_last_command != command) {
+		uint8_t cmd[2];
+		cmd[0] = static_cast<uint8_t>(command >> 8);
+		cmd[1] = static_cast<uint8_t>(command & 0xff);
+		_last_command = command;
+		return transfer(&cmd[0], 2, nullptr, 0);
+
+	} else {
+		return 0;
+	}
+}
+
+int SHT3X::read_data(uint16_t command, uint8_t *data_ptr, uint8_t length)
+{
+	set_pointer(command);
+
+	uint8_t raw_data[length];
+	transfer(nullptr, 0, &raw_data[0], length);
+
+	uint8_t crc_err = 0;
+
+	for (int i = 0; i < length / 3; ++i) {
+		uint8_t crc_data[2] = {raw_data[i * 3], raw_data[i * 3 + 1]};
+
+		if (raw_data[i * 3 + 2] != calc_crc(crc_data)) {
+			crc_err ++;
+		}
+
+		*(data_ptr + i * 2) = raw_data[i * 3];
+		*(data_ptr + i * 2 + 1) = raw_data[i * 3 + 1];
+	}
+
+	return crc_err;
+}
+
+int SHT3X::write_data(uint16_t command, uint8_t buffer[], uint8_t length)
+{
+	_last_command = command;
+
+	uint8_t cmd[2 + 3 * length / 2];
+	cmd[0] = static_cast<uint8_t>(command >> 8);
+	cmd[1] = static_cast<uint8_t>(command & 0xff);
+
+	for (int i = 0; i < length / 2; ++i) {
+		cmd[2 + 3 * i] = buffer[2 * i];
+		cmd[2 + 3 * i + 1] = buffer[2 * i + 1];
+		uint8_t crc_data[2] = {buffer[2 * i], buffer[2 * i + 1]};
+		cmd[2 + 3 * i + 2] = calc_crc(crc_data);
+	}
+
+	return transfer(&cmd[0], sizeof(cmd), nullptr, 0);
+}
+
+
+void SHT3X::sensor_compouse_msg(bool send)
+{
+	uint8_t data[4];
+	int error = read_data(SHT3x_CMD_FETCH_DATA, &data[0], 6);
+
+	if (error == PX4_OK) {
+		measurement_time = hrt_absolute_time();
+		measurement_index ++;
+
+		measured_temperature = (float) 175 * (data[0] << 8 | data[1]) / 65535 - 45;
+		measured_humidity = (float) 100 * (data[2] << 8 | data[3]) / 65535;
+
+		if (send) {
+			sensor_hygrometer_s msg{};
+			msg.timestamp = hrt_absolute_time();
+			msg.timestamp_sample = measurement_time;
+			msg.temperature = measured_temperature;
+			msg.humidity = measured_humidity;
+			msg.device_id = _sht_info.serial_number;
+			_sensor_hygrometer_pub.publish(msg);
+		}
+	}
+}
+
+
+int
+SHT3X::probe()
+{
+	uint8_t type[2];
+	uint8_t nvalid;
+
+	nvalid = read_data(SHT3X_CMD_READ_STATUS, &type[0], 3);
+	return (!(nvalid == PX4_OK));
+	// 0 means I can see sensor
+}
+
+int SHT3X::init()
+{
+	if (I2C::init() != PX4_OK) {
+		return PX4_ERROR;
+	}
+
+	_sensor_hygrometer_pub.advertise();
+	ScheduleOnInterval(50000);
+	return PX4_OK;
+}
+
+
+
+int SHT3X::init_sensor()
+{
+	set_pointer(SHT3X_CMD_CLEAR_STATUS);
+	px4_udelay(2000);
```

Comment:
```suggestion
	px4_usleep(2000);
```

First of all you probably shouldn't use udelay because it's a spinlock that prevents anything else on the system running for 2 milliseconds (potentially multiple IMU samples, etc). Using usleep is less precise (you'll get >= 2000 us), but actually lets the thread sleep and other things run. I would only use udelay if you need something very small (single digit microseconds).

Then in general with most of the drivers running in work queues we don't want them to sleep at all because it blocks the bus for other sensors. This is why most of the drivers now have the simple state machine structure so they can issue commands in one iteration, then schedule another run to come back and check the result.

So at a minimum you should change px4_udelay -> px4_usleep, but I would also encourage you to eliminate sleeps altogether to happily coexist on a shared bus.

https://github.com/PX4/PX4-Autopilot/blob/6d0339ba0c514b9a638373308c04d643f2bf6d5d/src/drivers/magnetometer/isentek/ist8310/IST8310.cpp#L104-L137

Reply:
Thank you for the explanation. This delay is required by the sensor and cannot be omitted. (Verified in datasheet). 

This concerns the initialization of the sensor, so these 2x2 ms sleeps are only performed at sensor start or error (for example, its disconnection and connection - Which should not occur under normal conditions). 

So I replaced it with a sleep procedure. But I will know about this when coding other things.